File: js/Protocol.js
const EventEmitter = require('events').EventEmitter;
const zmq = require('zeromq');

const terminator = 0;

const encode_label = str => Buffer.concat([Buffer.from(str, 'utf-8'), Buffer.from([terminator])]);

const decode_label = buf => {
	if (buf.length === 0 || buf[buf.length - 1] !== terminator) {
		return null;
	}
	return buf.subarray(0, buf.length - 1).toString('utf-8');
};

class Socket extends EventEmitter {
	constructor(config, logger = console) {
		super();
		this.logger = logger;
		this.host = config.host;
		const pub = zmq.socket('pub');
		const sub = zmq.socket('sub');
		this.pub = pub;
		this.sub = sub;
		this._init = [
			pub.connect(config.tx_url),
			sub.connect(config.rx_url)
		];
		sub.subscribe(encode_label(this.host));
		sub.on('message', (...parts) => this._recv(parts));
	}
	/* Returns void */
	async init() {
		await Promise.all(this._init);
	}
	/* Returns void */
	async send(envelope, ...parts) {
		const pub = this.pub;
		await pub.send(encode_label(envelope.remote), zmq.ZMQ_SNDMORE);
		await pub.send(encode_label(this.host), zmq.ZMQ_SNDMORE);
		await pub.send(encode_label(envelope.session), zmq.ZMQ_SNDMORE);
		await pub.send(encode_label(envelope.command), zmq.ZMQ_SNDMORE);
		for (let i = 0; i < parts.length; i++) {
			await pub.send(parts[i], i < parts.length - 1 ? zmq.ZMQ_SNDMORE : 0);
		}
	}
	/* Returns { envelope, parts } */
	_recv(msg) {
		if (msg.length < 4) {
			this.logger.warn('Insufficient message parts');
		}
		const [remote, host, session, command, ...parts] = msg;
		const envelope = new Envelope(
			decode_label(remote),
			decode_label(session),
			decode_label(command)
		);
		if (envelope.remote === null || envelope.session === null || envelope.command === null) {
			this.logger.warn('Invalid message envelope');
			return;
		}
		if (decode_label(host) !== this.host) {
			this.logger.warn('Received message addressed to some other host');
			return;
		}
		this.emit('message', envelope, ...parts);
	}
}

class Config {
	constructor(host = '', rx_url = '', tx_url = '', verbose = false) {
		this.host = host;
		this.rx_url = rx_url;
		this.tx_url = tx_url;
		this.verbose = verbose;
	}
}

class Envelope {
	constructor(remote = '', session = '', command = '') {
		if (arguments.length === 1 && typeof remote === 'object') {
			const o = remote;
			remote = o.remote;
			session = o.session;
			command = o.command;
		}
		this.remote = remote;
		this.session = session;
		this.command = command;
	}
}

Socket.Config = Config;
Socket.Envelope = Envelope;

module.exports = Socket;

File: python/Protocol.py
import asyncio
import zmq

class Config():
	rx_url = None
	tx_url = None
	remote = None
	hostname = None
	session = None
	timeout = None

class ZmqSocketError(RuntimeError):
	def __init__(self, msg):
		self.message = msg

class ReceiveTimeoutError(ZmqSocketError):
	pass

class InvalidMessageError(ZmqSocketError):
	pass

class OperationFailedError(ZmqSocketError):
	pass

class InvalidLabelError(ZmqSocketError):
	pass

class Envelope():
	remote = None
	session = None
	command = None

	def __init__(self, remote = None, session = None, command = None):
		self.remote = remote
		self.session = session
		self.command = command

terminator = 0

def encode_label(label):
	return bytes(label, 'utf-8') + bytes([terminator])

def decode_label(buf):
	if buf[-1] != terminator:
		raise InvalidLabelError('Invalid label')
	return str(buf[0:-1], 'utf-8')

class Socket():
	config = None
	ctx = None
	pub = None
	sub = None

	def __init__(self, ctx, config):
		self.config = config
		self.ctx = ctx
		self.pub = self.ctx.socket(zmq.PUB)
		self.sub = self.ctx.socket(zmq.SUB)
		self.pub.connect(config.tx_url)
		self.sub.connect(config.rx_url)
		self.sub.setsockopt(zmq.SUBSCRIBE, encode_label(config.hostname))

	def __enter__(self):
		self.ctx.__enter__();
		self.pub.__enter__();
		self.sub.__enter__();
		return self

	def __exit__(self, *args, **kwargs):
		self.sub.__exit__(*args, **kwargs);
		self.pub.__exit__(*args, **kwargs);
		self.ctx.__exit__(*args, **kwargs);
		return self

	async def send(self, envelope, *parts):
		msg = [
				encode_label(envelope.remote),
				encode_label(self.config.hostname),
				encode_label(envelope.session),
				encode_label(envelope.command),
		]
		msg += parts
		return await self.pub.send_multipart(msg)

	async def recv(self, timeout = None):
		if timeout is None:
			timeout = self.config.timeout
		if timeout is not None:
			timeout = timeout * 1000.0
		events = await self.sub.poll(timeout=timeout, flags=zmq.POLLIN)
		if not (events & zmq.POLLIN):
			raise ReceiveTimeoutError('Receive timed out')
		msg = await self.sub.recv_multipart()
		if len(msg) < 5:
			raise InvalidMessageError('Insufficient parts')
		try:
			local, remote, session, command = [decode_label(label) for label in msg[0:4]]
			if local != self.config.hostname:
				raise InvalidLabelError('Message is addressed to "' + local + '"')
		except InvalidLabelError as err:
			if timeout is None:
				return await recv(timeout)
			else:
				return ( None, None )
		parts = msg[4:]
		envelope = Envelope(remote, session, command)
		return ( envelope, parts )

Socket.Envelope = Envelope
